Make flutter test have a 3s timeout.

eseidelGoogle opened the pull request:
Fixes https://github.com/flutter/flutter/issues/2182

@Hixie @devoncarew @nex3

Commit:
Make flutter test have a 3s timeout.

Fixes https://github.com/flutter/flutter/issues/2182

@Hixie @devoncarew @nex3

Hixie commented on packages/flutter_tools/lib/src/test/flutter_platform.dart:
```diff
@@ -105,14 +106,33 @@ void main() {
+            print(line);
+        });
+      finalize();
+      throw new TimeoutException("Test took too long", _kPlatformTimeout);
```
Maybe "Test timed out. This may indicate a syntax error in the test (try using "flutter analyze" on the test). If an exception occurred, it will have been printed above this message."

(Or below this message, I forget in what order they print out.)

Hixie commented on packages/flutter_tools/lib/src/test/flutter_platform.dart:
```diff
@@ -105,14 +106,33 @@ void main() {
+      if (timeout != null) {
+        Timer timerToCancel = timeout;
+        timeout = null;
+        timerToCancel.cancel();
```
that's an odd way to do it. Why not (instead of the whole if block):
```dart
   timeout?.cancel();
   timeout = null;
```
If there's a reason to do it this way, please put a comment here explaining why.

Hixie commented:
Travis is showing failures for a lot of the tests. Maybe some bad interaction with how we run the tests in parallel? When are you starting the timer, relative to when a particular test starts running?

Ideally we wouldn't use a timer at all, and we'd immediately fail when the child process fails.

eseidelGoogle commented:
My approach appears to be wrong.  I attempted to add an exitCode listener to the process, but the process never exits on failure.  I also added print statements to the platform_plugin's handleDone methods on streams and those are never called in failure.  It looks like the child process is somehow swallowing the syntax errors and not reporting them to the parent.

I added some prints to package:test's vm's catchIsolateErrors() function, and it looks like we're not hitting that error handler in the syntax error case.  I suspect that it may due to how we've wired up our error handling for the isolate:
https://github.com/flutter/engine/blob/master/sky/engine/tonic/dart_error.cc#L15
Not sure yet.

eseidelGoogle commented:
I'm pretty sure the bug here is that Flutter's isolates not sending all expected errors through `Isolate.current.addErrorListener`, but I'm not quite sure what the fix is.  Likely will require understanding of how/when the Dart VM embedder is supposed to call error listeners.

I'm not quite sure how to write a minimal test of Flutter's Isolate.current.addErrorListener handling, but I suspect that we'd find that we don't call errorListeners for syntax errors.

FYI @johnmccutchan @nex3

eseidelGoogle commented:
So this change is wrong.  I believe the correct change is to make flutter pass  https://github.com/dart-lang/sdk/blob/master/tests/isolate/handle_error2_test.dart.  I'll update the original bug.
